### DISCUSSION TITLE
Mail soft failing due to "dh key too small"

### DISCUSSION THREAD

**Discussion by hameno** [postalserver/postal#1433] [Help with using Postal]
Trying to send mails to certain servers fail with the following error:

```
No SMTP servers were available for conntech.de. Tried mx01.goneo.de and mx02.goneo.de
SSL_connect returned=1 errno=0 state=error: dh key too small
```

This looks related to #878 but the provided patch did not work, I guess it has to be a missing setting in lib/postal/smtp_sender.rb

**Comment by willpower232** [comment]
Did you restart Postal after applying the patch from the PR?

**Comment by hameno** [comment]
Yes, postal was only started after this change (It is a Docker image with patched files)

**Reply by XenyFord** [reply]
Hi Sir,

I'm using CaddyServer and I have the same problem as shown in the screen below, and I made few changes to ` openssl.cnf ` by removing said lines. but still nothing worked for me. Can you help me on how to solve this please? 
![SSL_ERROR1](https://user-images.githubusercontent.com/89290310/134511370-1f282f02-edc1-4521-a30c-e623c277111f.jpg)

Thank you in advance!

**Comment by willpower232** [comment]
Oh hang on, is that error coming from Postal rather than your recipient?

Have you tried that `openssl` command against those email servers? It might be that Postal is having an issue trusting a lesser or more particular SSL setup.

**Comment by hameno** [comment]
It's coming from Postal when it tries to connect to external mail servers when trying to deliver the mails. I think it tries to setup a TLS connection but ultimately fails and never tried it without.

**Comment by hameno** [comment]
Tried the openssl command:
openssl s_client -connect mx01.goneo.de:25 -starttls smtp
```
CONNECTED(00000003)
depth=2 C = US, O = DigiCert Inc, OU = www.digicert.com, CN = DigiCert Global Root G2
verify return:1
depth=1 C = US, O = DigiCert Inc, OU = www.digicert.com, CN = RapidSSL TLS RSA CA G1
verify return:1
depth=0 CN = *.goneo.de
verify return:1
140497151870080:error:141A318A:SSL routines:tls_process_ske_dhe:dh key too small:../ssl/statem/statem_clnt.c:2150:
---
Certificate chain
 0 s:CN = *.goneo.de
   i:C = US, O = DigiCert Inc, OU = www.digicert.com, CN = RapidSSL TLS RSA CA G1
 1 s:C = US, O = DigiCert Inc, OU = www.digicert.com, CN = RapidSSL TLS RSA CA G1
   i:C = US, O = DigiCert Inc, OU = www.digicert.com, CN = DigiCert Global Root G2
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIGJzCCBQ+gAwIBAgIQCJJJqxz7H0Qz55XbsDLMpTANBgkqhkiG9w0BAQsFADBg
MQswCQYDVQQGEwJVUzEVMBMGA1UEChMMRGlnaUNlcnQgSW5jMRkwFwYDVQQLExB3
d3cuZGlnaWNlcnQuY29tMR8wHQYDVQQDExZSYXBpZFNTTCBUTFMgUlNBIENBIEcx
MB4XDTE4MDcxMjAwMDAwMFoXDTIwMDQwOTEyMDAwMFowFTETMBEGA1UEAwwKKi5n
b25lby5kZTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBALTOu4/ZvYhJ
UTASGuZuQATKU7RDGApaadTe1Qz+4V6WAhqZzMt7/8kPJ9MxvuBrCXBjDLCl+dON
dtZXR6ko2RGlN+qQqfeXGACOP1h+Qr6w+Bs1w4RIDVgXB3x4ooOTNyQ9xnGDFYsZ
Miwn85XFOYyW0MSjbvmkIPoSP+SBNZftcDk8JRQrKLAeNXfvcE0fXSvr28PNm5N3
YuWza9TMwEdj6NVO34wpsQw1n0eg8mfUEkHMlf0ED2QlmxMn9jg3wOxpk7Eivq/V
RxMcavMbvm75hYDRN/9QZyGU8TX4Crz1DCG5mkm9ImALqRQveN8B/0Mzt7mXDF57
Yh7iJr5R028CAwEAAaOCAyYwggMiMB8GA1UdIwQYMBaAFAzbbIJJD0pnCrgU7nrE
SFKI61Y4MB0GA1UdDgQWBBR6jH0xGgUfVeE/ZOD0neU+z1GOKjAfBgNVHREEGDAW
ggoqLmdvbmVvLmRlgghnb25lby5kZTAOBgNVHQ8BAf8EBAMCBaAwHQYDVR0lBBYw
FAYIKwYBBQUHAwEGCCsGAQUFBwMCMD8GA1UdHwQ4MDYwNKAyoDCGLmh0dHA6Ly9j
ZHAucmFwaWRzc2wuY29tL1JhcGlkU1NMVExTUlNBQ0FHMS5jcmwwTAYDVR0gBEUw
QzA3BglghkgBhv1sAQIwKjAoBggrBgEFBQcCARYcaHR0cHM6Ly93d3cuZGlnaWNl
cnQuY29tL0NQUzAIBgZngQwBAgEwdgYIKwYBBQUHAQEEajBoMCYGCCsGAQUFBzAB
hhpodHRwOi8vc3RhdHVzLnJhcGlkc3NsLmNvbTA+BggrBgEFBQcwAoYyaHR0cDov
L2NhY2VydHMucmFwaWRzc2wuY29tL1JhcGlkU1NMVExTUlNBQ0FHMS5jcnQwCQYD
VR0TBAIwADCCAXwGCisGAQQB1nkCBAIEggFsBIIBaAFmAHUApLkJkLQYWBSHuxOi
zGdwCjw1mAT5G9+443fNDsgN3BAAAAFkjZl2cgAABAMARjBEAiBozISp4dAAqxnD
QDr+nAtcAQG6kV3WaK9bxuX/PZqOiAIgPgUVfNf7vOGFiaae8XcMiNhGx803x8mP
zAsecw7s45QAdgCHdb/nWXz4jEOZX73zbv9WjUdWNv9KtWDBtOr/XqCDDwAAAWSN
mXdAAAAEAwBHMEUCIQC/ELzGDN/RlOL5k75YKSr+dqqHA6nJ25D5ViPyDTl+kAIg
WA1BVO8bojd8Cdkz9phZmMY8ol1hdwSSIT0CFJL6ho4AdQC72d+8H4pxtZOUI5eq
kntHOFeVCqtS6BqQlmQ2jh7RhQAAAWSNmXabAAAEAwBGMEQCIE9ply0GxnPALCx8
7Eaxr/P1vs9VK2KFHLQo6gqbtMKBAiBzzoLEk5Y5B6egZbJWU+NdC708d2MktFUJ
Wd6P8aAkizANBgkqhkiG9w0BAQsFAAOCAQEAqlYYstVaT6bzDSAiHNLOvvUU1V3l
WcCDazZqUjBKKRSAe6JA45GHBxALwFGaK+jFFpMGpgDskkd1Ya2bGaoW8qBGghPA
ALPsAgU+aECGUOmtAkdTj5CajVa6XgyT0/GB2oDb58RuFNCQnk5dD5p+pQrrLZWO
m9pXPrU4xaapHnvOJc53PX+m+YqXraCiDsZRH5SvMlsaplJSF67Vm3fFzxgg+lny
Ls4w6irVozpht08siBBua9IRQvsv4xhDcNs+x/i3antlscmfZurh97IDoMIhoW2c
iwbaGK67rl1aMoPw+brJ3Td/s10cEr/fkzyQ/Pel/oYpScCvMHYYnqNqAg==
-----END CERTIFICATE-----
subject=CN = *.goneo.de

issuer=C = US, O = DigiCert Inc, OU = www.digicert.com, CN = RapidSSL TLS RSA CA G1

---
No client certificate CA names sent
---
SSL handshake has read 3573 bytes and written 345 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : 0000
    Session-ID: 
    Session-ID-ctx: 
    Master-Key: 
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1576740102
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
---
```
Also shows an error about this, but shouldn't postal then switch to not using TLS?

**Comment by willpower232** [comment]
Interestingly I get a different output without the error if I run the same command. What is the output of `openssl version`?

**Comment by hameno** [comment]
```OpenSSL 1.1.1d  10 Sep 2019```

The Docker image is based on `ruby:2.6`

**Comment by hameno** [comment]
Found a workaround for the openssl command in https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=907788
Removed the line and will now deploy a new version without this config

**Comment by hameno** [comment]
Workaround verified, mail is now delivered to this mail server

**Reply by XenyFord** [reply]
Hi Hameno,
Could you please let me know what exact changes you made to make it worked? and where? Thank you in advance!

**Comment by willpower232** [comment]
Thanks for the extra research. FYI Postal is currently only fully supported on Ruby 2.3 but hopefully the others should be updating the gems to improve support for more recent Ruby versions.

**Reply by XenyFord** [reply]
Hi Mr. Will,

I'm still getting the same error as below though I tried with many changes like modifying ` Openssl.cnf ` and etc as suggested.

` No SMTP servers were available for vip.sina.com. Tried mx3.vip.sina.com, mx1.vip.sina.com, mx2.vip.sina.com, and sinamx.vip.sina.com `
`  SSL_connect returned=1 errno=0 state=error: dh key too small `
` Support Ref: VHFZOSSZ `

Please let me know If there is solution for this. Thank you!

**Comment by vishnus** [comment]
Adding to this thread, incase anyone is facing this issue in Ubuntu 20.04 and come across this thread..  please go through https://askubuntu.com/questions/1233186/ubuntu-20-04-how-to-set-lower-ssl-security-level
